lib/syskit/network_generation/merge_solver.rb

```diff
@@ -220,16 +216,49 @@ def may_merge_task_contexts?(merged_task, task)
                     return false
                 end
 
+                true
+            end
+
+            def merge_task_contexts_with_same_agent(enabled:)
+                @merge_when_identical_agents = enabled
+                r = yield
+                @merge_when_identical_agents = false
+
+                r
+            end
+
+            # Tests whether task.merge(target_task) is a valid operation
+            #
+            # @param [Syskit::TaskContext] task
+            # @param [Syskit::TaskContext] target_task
+            #
+            # @return [false,true] if false, the merge is not possible. If
+            #   true, it is possible.
+            def may_merge_task_contexts?(merged_task, task)
+                return false unless may_merge_components?(merged_task, task)
+
                 # Merges involving a deployed task can only involve a
                 # non-deployed task as well
-                if task.execution_agent && merged_task.execution_agent
+                unless mergeable_agents?(merged_task, task)
                     info "rejected: deployment attribute mismatches"
                     return false
                 end
 
                 true
             end
 
+            def mergeable_agents?(merged_task, task)
+                unless merge_when_identical_agents?
+                    return true unless task.execution_agent && merged_task.execution_agent
+
+                    return false
```

Comment:
This is `return !(task.execution_agent && merged_task.execution_agent)`, which is a lot easier to parse.